doc/kernel/other/memory_domain.rst

```diff
@@ -0,0 +1,192 @@
+    u8_t __aligned(32) app0_buf[32];
+    u8_t __aligned(32) app1_buf[32];
+
+    K_MEM_PARTITION_DEFINE(app0_part0, app0_buf, sizeof(app0_buf),
+                           K_MEM_PARTITION_P_RW_U_RW);
+
+    K_MEM_PARTITION_DEFINE(app0_part1, app1_buf, sizeof(app1_buf),
+                           K_MEM_PARTITION_P_RW_U_RO);
+
+    struct k_mem_partition *app0_parts[] = {
+        app0_part0,
+        app0_part1
+    };
+
+    k_mem_domain_init(&app0_domain, "app0_domain",
+                      ARRAY_SIZE(app0_parts), app0_parts);
+
+The following code shows how to add memory partitions into a initialized
+memory domain one by one.
+
+.. code-block:: c
+
+    /* the start address of the MPU region needs to align with its size */
+    u8_t __aligned(32) app0_buf[32];
+    u8_t __aligned(32) app1_buf[32];
+
+    K_MEM_PARTITION_DEFINE(app0_part0, app0_buf, sizeof(app0_buf),
+                           K_MEM_PARTITION_P_RW_U_RW);
+
+    K_MEM_PARTITION_DEFINE(app0_part1, app1_buf, sizeof(app1_buf),
+                           K_MEM_PARTITION_P_RW_U_RO);
+
+    k_mem_domain_add_partition(&app0_domain, &app0_part0);
+    k_mem_domain_add_partition(&app0_domain, &app0_part1);
+
+.. note::
+    The maximum number of memory partitions would be limited by maixmum
+    number of MPU regions or maximum number of MMU tables.
+
+Add Threads into a Memory Domain
+================================
+
+Adding threads into a memory domain grants threads to access the memory
+partitions in the memory domain.
+
+The following code shows how to add threads into a memory domain.
+
+.. code-block:: c
+
+    k_mem_domain_add_thread(&app0_domain, &app_thread_id);
+
+Remove a Memory Partition from a Memory Domain
+==============================================
+
+The following code shows how to remove a memory partition from a memory
+domain.
+
+.. code-block:: c
+
+    k_mem_domain_remove_partition(&app0_domain, &app0_part1);
+
+k_mem_domain_remove_partition() API would find the memory partition
+that matches the given parameter and remove it from the memory domain.
+
+.. note::
+    Once a memory partition was removed from the memory domain, threads
+    in the memory domain lose the access permission to the removed memory
+    partition immediately.
+
+Remove a Thread from the Memory Domain
+======================================
+
+The following code shows how to remove a thread from the memory domain.
+
+.. code-block:: c
+
+    k_mem_domain_remove_thread(&app0_domain, &app_thread_id);
+
+.. note::
+    Once a thread was removed from the memory domain, this thread loses
+    the access permission to the memory partitions in the memory domain
+    that this thread has belonged to immediately.
+
+Destroy a Memory Domain
+=======================
+
+The following code shows how to destroy a memory domain.
+
+.. code-block:: c
+
+    k_mem_domain_destroy(&app0_domain);
+
+.. note::
+    Once a memory domain was destroyed, threads in the destroyed memory
```

Comment:
How about:

     Threads will immediately lose access permission to the
     memory partitions in the destroyed memory domain.